File: src/csharp/Foundgine.Core/Semantic/Metadata/MetadataRegistry.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>
/// In-memory registry of the static metadata used by resolution and planning.
/// A future AOT generator can emit an implementation with the same contract.
/// </summary>
public sealed class MetadataRegistry : IMetadataCatalog, IMutationSchema
{
    private readonly Dictionary<EntityId, EntityMetadata> _entities = new();

    private readonly Dictionary<RelationshipId, RelationshipMetadata> _relationships = new();

    private readonly Dictionary<ModelId, ModelMetadata> _models = new();

    private readonly Dictionary<ConnectionId, ConnectionMetadata> _connections = new();

    private readonly List<ConversionMetadata> _conversions = new();

    private readonly Dictionary<AuthorizationId, AuthorizationMetadata> _authorizations = new();

    public IEnumerable<EntityMetadata> Entities =>
        _entities.Values;

    public IEnumerable<RelationshipMetadata> Relationships =>
        _relationships.Values;

    public IEnumerable<ModelMetadata> Models =>
        _models.Values;

    public IEnumerable<ConnectionMetadata> Connections =>
        _connections.Values;

    public IEnumerable<ConversionMetadata> Conversions =>
        _conversions;

    public IEnumerable<AuthorizationMetadata> Authorizations =>
        _authorizations.Values;

    public void Register(EntityMetadata metadata)
    {
        ArgumentNullException.ThrowIfNull(metadata);

        _entities[metadata.EntityId] = metadata;
    }

    public void Register(RelationshipMetadata relationship)
    {
        ArgumentNullException.ThrowIfNull(relationship);

        _relationships[relationship.Id] = relationship;
    }

    public void Register(ModelMetadata model)
    {
        ArgumentNullException.ThrowIfNull(model);

        _models[model.Id] = model;
    }

    public void Register(ConnectionMetadata connection)
    {
        ArgumentNullException.ThrowIfNull(connection);

        _connections[connection.Id] = connection;
    }

    public void Register(AuthorizationMetadata authorization)
    {
        ArgumentNullException.ThrowIfNull(authorization);

        _authorizations[authorization.Id] = authorization;
    }

    public AuthorizationMetadata GetAuthorization(
        AuthorizationId id)
    {
        return _authorizations.TryGetValue(
            id,
            out var authorization)
            ? authorization
            : throw new KeyNotFoundException(
                $"Authorization {id} was not registered.");
    }

    public void Register(ConversionMetadata conversion)
    {
        ArgumentNullException.ThrowIfNull(conversion);

        if (_conversions.Any(x =>
                x.SourceType == conversion.SourceType &&
                x.TargetType == conversion.TargetType))
        {
            throw new InvalidOperationException(
                $"Duplicate Foundgine conversion " +
                $"{conversion.SourceType} -> {conversion.TargetType}.");
        }

        _conversions.Add(conversion);
    }

    public bool TryGet(
        EntityId id,
        out EntityMetadata metadata)
    {
        return _entities.TryGetValue(
            id,
            out metadata!);
    }

    public EntityMetadata Get(
        EntityId id)
    {
        return _entities.TryGetValue(
            id,
            out var metadata)
            ? metadata
            : throw new KeyNotFoundException(
                $"Entity {id} was not registered.");
    }

    public EntityMetadata GetEntity(
        EntityId entityId)
    {
        return Get(entityId);
    }

    public bool TryGet(
        ModelId id,
        out ModelMetadata model)
    {
        return _models.TryGetValue(
            id,
            out model!);
    }

    public ModelMetadata GetModel(
        ModelId id)
    {
        return Get(id);
    }

    public ModelMetadata Get(
        ModelId id)
    {
        return _models.TryGetValue(
            id,
            out var model)
            ? model
            : throw new KeyNotFoundException(
                $"Model {id} was not registered.");
    }

    public bool TryGet(
        ConnectionId id,
        out ConnectionMetadata connection)
    {
        return _connections.TryGetValue(
            id,
            out connection!);
    }

    public ConnectionMetadata GetConnection(
        ConnectionId id)
    {
        return Get(id);
    }

    public ConversionMetadata? FindConversion(
        Type sourceType,
        Type targetType)
    {
        return _conversions.FirstOrDefault(x =>
            x.SourceType == sourceType &&
            x.TargetType == targetType);
    }

    public ConnectionMetadata Get(
        ConnectionId id)
    {
        return _connections.TryGetValue(
            id,
            out var connection)
            ? connection
            : throw new KeyNotFoundException(
                $"Connection {id} was not registered.");
    }

    public bool TryGet(
        RelationshipId id,
        out RelationshipMetadata relationship)
    {
        return _relationships.TryGetValue(
            id,
            out relationship!);
    }

    public RelationshipMetadata Get(
        RelationshipId id)
    {
        return _relationships.TryGetValue(
            id,
            out var relationship)
            ? relationship
            : throw new KeyNotFoundException(
                $"Relationship {id} was not registered.");
    }

    public RelationshipMetadata GetRelationship(
        RelationshipId relationshipId)
    {
        return Get(relationshipId);
    }

    public MutationEntitySchema GetEntitySchema(
        EntityId entityId)
    {
        var entity = Get(entityId);

        var fields = entity.EffectiveFields.ToDictionary(
            field => field.Id,
            field => field.Column?.ColumnId);

        var columns = entity.Columns
            .Select(column => column.Id)
            .ToHashSet();

        return new MutationEntitySchema(
            entity.EntityId,
            entity.Name,
            columns,
            fields,
            entity.PrimaryKey?.ColumnId);
    }

    MutationEntitySchema IMutationSchema.GetEntity(
        EntityId entityId)
    {
        return GetEntitySchema(entityId);
    }

    public MutationRelationshipSchema GetRelationshipSchema(
        RelationshipId relationshipId)
    {
        var relationship = Get(relationshipId);

        return new MutationRelationshipSchema(
            relationship.Id,
            relationship.Source,
            relationship.Target,
            relationship.Name,
            relationship.SourceKey.ColumnId,
            relationship.TargetKey.ColumnId);
    }

    MutationRelationshipSchema IMutationSchema.GetRelationship(
        RelationshipId relationshipId)
    {
        return GetRelationshipSchema(relationshipId);
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/MetadataRegistry.java
package com.foundgine.core.semantic.metadata;

import com.foundgine.core.abstractions.*;
import java.util.*;

/** In-memory registry of static metadata. */
public final class MetadataRegistry implements IMetadataCatalog {
	private final Map<EntityId, EntityMetadata> entities = new LinkedHashMap<>();
	private final Map<RelationshipId, RelationshipMetadata> relationships = new LinkedHashMap<>();
	private final Map<ModelId, ModelMetadata> models = new LinkedHashMap<>();
	private final Map<ConnectionId, ConnectionMetadata> connections = new LinkedHashMap<>();
	private final List<ConversionMetadata> conversions = new ArrayList<>();
	private final Map<AuthorizationId, AuthorizationMetadata> authorizations = new LinkedHashMap<>();

	@Override
	public Iterable<EntityMetadata> entities() {
		return Collections.unmodifiableCollection(entities.values());
	}

	@Override
	public Iterable<RelationshipMetadata> relationships() {
		return Collections.unmodifiableCollection(relationships.values());
	}

	public Iterable<ModelMetadata> models() {
		return Collections.unmodifiableCollection(models.values());
	}

	public Iterable<ConnectionMetadata> connections() {
		return Collections.unmodifiableCollection(connections.values());
	}

	public Iterable<ConversionMetadata> conversions() {
		return Collections.unmodifiableList(conversions);
	}

	public Iterable<AuthorizationMetadata> authorizations() {
		return Collections.unmodifiableCollection(authorizations.values());
	}

	public void register(EntityMetadata metadata) {
		require(metadata);
		entities.put(metadata.entityId(), metadata);
	}

	public void register(RelationshipMetadata metadata) {
		require(metadata);
		relationships.put(metadata.id(), metadata);
	}

	public void register(ModelMetadata metadata) {
		require(metadata);
		models.put(metadata.id(), metadata);
	}

	public void register(ConnectionMetadata metadata) {
		require(metadata);
		connections.put(metadata.id(), metadata);
	}

	public void register(AuthorizationMetadata metadata) {
		require(metadata);
		authorizations.put(metadata.id(), metadata);
	}

	public void register(ConversionMetadata conversion) {
		require(conversion);
		if (conversions.stream()
				.anyMatch(x -> x.sourceType() == conversion.sourceType() && x.targetType() == conversion.targetType()))
			throw new IllegalStateException("Duplicate Foundgine conversion " + conversion.sourceType() + " -> "
					+ conversion.targetType() + ".");
		conversions.add(conversion);
	}

	public boolean tryGet(EntityId id) {
		return entities.containsKey(id);
	}

	public EntityMetadata get(EntityId id) {
		return requireGet(entities, id, "Entity");
	}

	@Override
	public EntityMetadata getEntity(EntityId id) {
		return get(id);
	}

	public boolean tryGet(ModelId id) {
		return models.containsKey(id);
	}

	public ModelMetadata get(ModelId id) {
		return requireGet(models, id, "Model");
	}

	@Override
	public ModelMetadata getModel(ModelId id) {
		return get(id);
	}

	public boolean tryGet(ConnectionId id) {
		return connections.containsKey(id);
	}

	public ConnectionMetadata get(ConnectionId id) {
		return requireGet(connections, id, "Connection");
	}

	@Override
	public ConnectionMetadata getConnection(ConnectionId id) {
		return get(id);
	}

	public boolean tryGet(RelationshipId id) {
		return relationships.containsKey(id);
	}

	public RelationshipMetadata get(RelationshipId id) {
		return requireGet(relationships, id, "Relationship");
	}

	@Override
	public RelationshipMetadata getRelationship(RelationshipId id) {
		return get(id);
	}

	@Override
	public AuthorizationMetadata getAuthorization(AuthorizationId id) {
		return requireGet(authorizations, id, "Authorization");
	}

	@Override
	public ConversionMetadata findConversion(Class<?> sourceType, Class<?> targetType) {
		return conversions.stream().filter(x -> x.sourceType() == sourceType && x.targetType() == targetType)
				.findFirst().orElse(null);
	}

	private static void require(Object value) {
		Objects.requireNonNull(value);
	}

	private static <K, V> V requireGet(Map<K, V> map, K key, String kind) {
		V value = map.get(key);
		if (value == null)
			throw new NoSuchElementException(kind + " " + key + " was not registered.");
		return value;
	}
}
